libraries/AP_Generator/AP_Generator_Loweheiser.cpp

```diff
@@ -0,0 +1,832 @@
+            }
+        }
+        last_packet_received_ms = now_ms;
+        good_message = true;
+        break;
+    }
+    case MAVLINK_MSG_ID_COMMAND_ACK:
+        // only accept acks once we've seen a loweheiser packet:
+        if (!seen_good_message) {
+            return;
+        }
+        mavlink_msg_command_ack_decode(&msg, &ack_packet);
+        last_ack_packet_ms = AP_HAL::millis();
+        break;
+    }
+
+    if (!good_message) {
+        return;
+    }
+    if (seen_good_message) {
+        return;
+    }
+    seen_good_message = true;
+
+    mavlink_channel = &channel;
+    sysid = msg.sysid;
+    compid = msg.compid;
+}
+
+// send mavlink generator status
+void AP_Generator_Loweheiser::send_generator_status(const GCS_MAVLINK &channel)
+{
+    if (last_packet_received_ms == 0) {
+        // nothing to report
+        return;
+    }
+
+    uint64_t status = 0;
+    switch (commanded_runstate) {
+    case RunState::STOP:
+        // we checking RPM here but confounded by starter motor running....
+        status |= MAV_GENERATOR_STATUS_FLAG_OFF;
+        break;
+    case RunState::COOLING_DOWN:
+    case RunState::IDLE:
+        if (generator_ok_to_run()) {
+            status |= MAV_GENERATOR_STATUS_FLAG_READY;
+        }
+        break;
+    case RunState::RUN:
+        status |= MAV_GENERATOR_STATUS_FLAG_GENERATING;
+        break;
+    case RunState::WARMING_UP:
+        status |= MAV_GENERATOR_STATUS_FLAG_WARMING_UP;
+        break;
+    }
+
+    // note that the EFI unit can be turned off on the Loweheiser, in
+    // which case we receive NaN values back.  We translate these into
+    // MAVLink "unknown" values throughout here:
+
+    if (!isnan(packet.efi_clt) &&
+        packet.efi_clt > temp_for_overtemp_warning) {
+        status |= MAV_GENERATOR_STATUS_FLAG_OVERTEMP_WARNING;
+    }
+
+    int16_t rectifier_temp = INT16_MAX;
+    if (!isnan(packet.rectifier_temp)) {
+        rectifier_temp = packet.rectifier_temp;
+    }
+    int16_t generator_temp = INT16_MAX;
+    if (!isnan(packet.generator_temp)) {
+        generator_temp = packet.generator_temp;
+    }
+
+    // not all loweheiser firmwares provide runtime and
+    // time-until-maintenance, so provide values we have in our
+    // parameters for those:
+    uint32_t runtime = packet.runtime;
+    if (runtime == UINT32_MAX) {
+        runtime = total_runtime + runtime_delta_ms/1000;
+    }
+    uint32_t maint_time = packet.until_maintenance;
+    if (maint_time == INT32_MAX) {
+        maint_time = time_until_maintenance - runtime_delta_ms/1000;
+    }
+
+    mavlink_msg_generator_status_send(
+        channel.get_chan(),
+        status,
+        _rpm, // generator_speed
+        packet.curr_batt,  // Current into/out of battery
+        packet.curr_gen, // load_current; Current going to UAV
+        std::numeric_limits<double>::quiet_NaN(), // power_generated; the power being generated
```

Comment:
according to the loweheiser spec. `curr_rot` is the total consumed current by the system? Thus this suggestion.

But we should check they aren't blindly doing `curr_rot = curr_batt + cur_gen` 

As we really only want the charging battery current `curr_rot = curr_batt_charging + cur_gen` 

```suggestion
        (packet.curr_rot * _voltage),               // power_generated (w)
```